sycl/source/detail/platform_impl.cpp

```diff
@@ -338,6 +341,10 @@ static void filterDeviceFilter(vector_class<RT::PiDevice> &PiDevices,
     DeviceNum++;
   }
   PiDevices.resize(InsertIDx);
+  // remember the last backend that has gone through this filter function
+  // to assign a unique device id number across platforms that belong to
+  // the same backend. For example, opencl:cpu:0, opencl:acc:1, opencl:gpu:2
+  Plugin->setLastDeviceId(DeviceNum);
```

Comment:
I am not quite sure but it may happen that when I call platform::get_devices() twice in the same program I can get different device_nums for the second call.

Reply:
Correct.
We need to assign different device numbers across different platforms within the same backend.
So, device::get_devices() will reset the LastDeviceId to zero when it iterates all plugins.
But, if platform::get_devices() is called twice for the same platform, it will receive different device ids.
It seems I have to maintain LastDeviceId per platform, and plugin->getLastDeviceId(platform) should return the lastDeviceId of the previously seen platform.
For example, let's assume opencl has 3 platforms: acc, cpu, and gpu platforms.
opencl plugin will first initialize LastDeviceIds[acc] = 0, LastDeviceIds[cpu] = 0, LastDeviceIds[gpu] = 0.
And whenever, getLastDevice(platform) is called, it returns lastDeviceIds[platformId - 1].
Then, platform::get_devices() will always give the same device id no matter how many times it is called.

Reply:
Done. Thanks.